ocaml/parsing/ast_invariants.ml

```diff
@@ -45,9 +49,31 @@ let iterator =
     | Ptype_record [] -> empty_record loc
     | _ -> ()
   in
+  let jtyp _self loc (jty : Jane_syntax.Core_type.t) =
+    match jty with
+    | Jtyp_local (Ltyp_local _) -> misplaced_local loc
+  in
+  let typ_allow_local ty =
+    match Jane_syntax.Core_type.of_ast ty with
+    | Some (Jtyp_local (Ltyp_local ty), []) -> ty
+    | None -> ty
+    | Some (Jtyp_local (Ltyp_local _), _::_) ->
+        local_with_attributes ty.ptyp_loc
+  in
   let typ self ty =
-    super.typ self ty;
+    begin match ty.ptyp_desc with
+    | Ptyp_arrow (lab, ty1, ty2) ->
+        let without_locals =
+          Ptyp_arrow(lab, typ_allow_local ty1, typ_allow_local ty2)
+        in
+        super.typ self { ty with ptyp_desc = without_locals }
+    | _ ->
+        super.typ self ty
+    end;
```

Comment:
I think that's exactly why it goes here.  The goal is, if you have `local_ a -> local_ b`, to never recurse directly into `local_ a` or `local_ b`.  `super.typ` handles the recursion; thus, to avoid seeing `local_ foo`, we strip them before recursing.  We should maybe do a jane-syntax check here _too_, but that's separate from the below.  See also the "allow unary tuple" case on lines 83--90.

Reply:
I agree that we should do a jane syntax check here too.